ldap/ldap.go

```diff
@@ -0,0 +1,127 @@
+)
+
+//ValidateUser validates user by extracting creds from the request body, then
+//    authenticating against an ID backend and filling out the User struct therefrom
+func ValidateUser(req *http.Request, authConfig config.AuthConfig) (User, error) {
+
+	// access LDAP as the user we are validating
+	username, pwd, _ := req.BasicAuth()
+
+	log.Debugf("Connecting to LDAP with config: %v", authConfig)
+	bindDN := strings.Replace(authConfig.LDAP.LDAPBindDN, "{USERID}", username, -1)
+	log.Debugf("bindDN : %s", bindDN)
+
+	ldapAttributes := []string{authConfig.LDAP.LDAPMemberOfKey, authConfig.LDAP.LDAPUserIDKey, authConfig.LDAP.LDAPMailIDKey}
+	client := ldap.LDAPClient{
+		Base:               authConfig.LDAP.LDAPBase,
+		Host:               authConfig.LDAP.LDAPHost,
+		Port:               authConfig.LDAP.LDAPPort,
+		UseSSL:             authConfig.LDAP.LDAPUseSSL,
+		InsecureSkipVerify: authConfig.LDAP.LDAPInsecureSkipVerify,
+		BindDN:             bindDN,
+		BindPassword:       pwd,
+		UserFilter:         authConfig.LDAP.LDAPUserFilter,
+		GroupFilter:        authConfig.LDAP.LDAPGroupFilter,
+		Attributes:         ldapAttributes,
+	}
+	defer client.Close()
+
+	// uname and password (ID) correct?
+	authOk, err := isAuthenticated(&client, username, pwd)
+
+	if authOk { // get the rest of the info for this person -- groups, app ownershop etc.
+		user, err := getUser(&client, username, strings.ToLower(strings.TrimSpace(authConfig.LDAP.LDAPRoleFilter)), authConfig)
+		return user, err
+	}
+	log.Debugf("Authentication failed with LDAP : %v", err)
+	return User{}, err
+}
+
+// isAuthenticated checks whether uname and password are correct according to ID (LDAP currently) backend
+func isAuthenticated(ldapClient *ldap.LDAPClient, username, password string) (bool, error) {
+	ok, _, err := ldapClient.Authenticate(username, password)
+	if err != nil {
+		log.Errorf("Error authenticating user %s: %+v", username, err)
+		return false, err
+	}
+	if !ok {
+		log.Errorf("Authenticating failed for user %s", username)
+		return false, err
+	}
+	return true, nil
+}
+
+// getUser takes a structure with only the bare creds filled in and consults the roles backend (LDAP) to return
+// a completely filled out structure with uid, username, groups and app ownership information
+func getUser(ldapClient *ldap.LDAPClient, username, roleFilter string, authConfig config.AuthConfig) (User, error) {
+	groups := []string{}
+	var user User
+
+	ldapAttributes := []string{authConfig.LDAP.LDAPMemberOfKey, authConfig.LDAP.LDAPUserIDKey, authConfig.LDAP.LDAPMailIDKey}
+	searchFilter := strings.Replace(authConfig.LDAP.LDAPSearchFilter, "{USERID}", username, -1)
+	searchRequest := ldapv2.NewSearchRequest(
+		authConfig.LDAP.LDAPBase,
+		ldapv2.ScopeWholeSubtree,
+		ldapv2.NeverDerefAliases, 0, 0, false,
+		searchFilter,
+		ldapAttributes,
+		nil,
+	)
+	sr, err := ldapClient.Conn.Search(searchRequest)
+	if err != nil {
+		log.Errorf("Error searching AD : %s", err.Error())
```

Comment:
This isn't (hopefully isn't) ActiveDirectory specific, so I would change this message to say "Error searching LDAP", not "AD":

```suggestion
		log.Errorf("Error searching LDAP: %s", err.Error())
```

Reply:
Changed the log comment